enhancements/update/automatic-updates.md

```diff
@@ -0,0 +1,166 @@
+This enhancement would provide those administrators with a convenient property instead.
+
+### Goals
+
+* Make it easy to opt in to and out of automatic updates.
+
+### Non-Goals
+
+* Support calendar gating or other logic about when updates may be attempted.
+* Choose whether the installer defaults to enabling automatic updates or not for new clusters.
+
+## Proposal
+
+As proposed in [this API pull request][api-pull-request], to add a new property:
+
+```go
+// automaticUpdates enables automatic updates.
+// +optional
+AutomaticUpdates bool `json:"automaticUpdates,omitempty"`
+```
+
+to [`ClusterVersionSpec`][api-spec] to record the administrator's preference.
+
+### User Stories
+
+#### Canary Clusters
+
+Alice has a canary clusters she wants to update as soon as an updated release is available in her configured channel.
+Before this enhancement she would need to construct a poller to run `oc adm upgrade --to-latest` or similar.
+With this enhancement, she can set `automaticUpdates` and does not need to write an external poller.
+
+### Risks and Mitigations
+
+Cluster-managed updates are a relatively new thing, and while we have dozens of OpenShift 4 releases out so far, we have had a few updates result in stuck clusters (like [this][machine-config-operator-drain-wedge].
+Allowing automatic updates would make it more likely that upgrade failures happen when there is no admistrator actively watching to notice and recover from the failure.
+This is mitigated by:
+
+* Alerting.
+    The cluster should fire alerts (FIXME: which ones?) when an upgrade gets into trouble.
+    Adminstrators can configure the cluster to push those alerts out to the on-call admistrator to recover the cluster.
+* Stability testing.
+    We are continually refining our CI suite and processing Telemetry from live clusters in order to assess the stability of each upgrade.
+    We will not place upgrades in production channels unless they have proven themselves stable in earlier testing, and we will remove upgrades from production channels if a live cluster trips over a corner case that we do not yet cover in pretesting.
+
+There are also potential future mitigations:
+
+* Phased releases, where Cincinnati spreads an update suggestion out over a configurable time window.
+    With hard cutovers and automatic updates, many clusters in a given channel could attempt a new update simultaneously.
+    If that update proves unstable, many of those upgrades would already be in progress by the time the first Telemetry comes back with failure messages.
+    A phased rollout would limit the number of simultaneously updating clusters to give Telemetry time to come back so we could stop recommending upgrades that proved unstable on live clusters not yet covered in pretesting.
+
+There is also a security risk where a compromised upstream Cincinnati could recommend cluster updates that were not in the cluster's best interest (e.g. 4.2.4 -> 4.1.0).
```

Comment:
Not sure if relevant for OpenShift mitigations, but Zincati has client-side checks and knobs to prevent auto-downgrades: https://github.com/coreos/zincati/blob/dbb0b0a8884435f2b1186b2228199cd4adb6f705/docs/usage/auto-updates.md#updates-ordering-and-downgrades

Reply:
> Not sure if relevant for OpenShift mitigations...

We could grow this.  But sometimes you might *want* to recommend a downgrade.  E.g, 4.y.7 turns out to explode after 24h because of broken cert rotation, and a fixed 4.y.8 is 48h out, so you recommend 4.y.7->4.y.6 until then to get folks back to a safe place.

Reply:
I think it makes sense to version updates separately from the cluster versions, e.g.
 ```
update-1.0 => (ocp-4.y.6 -> ocp-4.y.7)
update-1.1 => (ocp-4.y.7 -> ocp-4.y.6)
update-1.2 => (ocp-4.y.6 -> ocp-4.y.8, ocp-4.y.7 -> ocp-4.y.8)
```
You always want the latest update version, which may or may not upgrade you to the highest cluster version.

Another way to address rollback attacks, as well as freeze attacks: TUF uses timestamps to certify updates for short durations, requiring constant refreshing to ensure the latest update metadata. Timestamp ordering is simple enough but does require secure and accurate time sources (for TUF, see the timestamp.json section: [tuf-spec.md#4-document-formats](https://github.com/theupdateframework/specification/blob/master/tuf-spec.md#4-document-formats)

Reply:
For [libostree](https://github.com/ostreedev/ostree/) the timestamp in the commit is covered by the GPG signature; we don't do anything about comparing with the system wall clock, just require that the timestamp in the new commit increases.

The TUF threat model helps in a DoS attack where a MITM attacker just tells you there are no more updates.  This is useful, but comes with a lot of overhead.

Reply:
For OpenShift though the MCO ignores that bit, and obviously the libostree part isn't use for the rest of the container images anyways.  

So I guess I'm just arguing that "signed timestamps" work pretty well and would likely be easy to add to the CVO if it doesn't do it today.

Reply:
> You always want the latest update version, which may or may not upgrade you to the highest cluster version.

We've talked about signing update recommendations, but you'd probably need to sign each of them separately.  E.g.

```yaml
from: 4.3.0
to: 4.3.1
expires: 2020-02-08T00:00Z
```

That's possible, but would be a fair bit of work to put in.  Protections like "require admin overrides before applying downgrades" are coarser (and as above, sometimes you want downgrades), but would be easier to implement in the short term.

> So I guess I'm just arguing that "signed timestamps" work pretty well and would likely be easy to add to the CVO if it doesn't do it today.

For the releases themselves, we have a version number and creation timestamp, both of which are covered by the image signature.  Sorting on version number seems more sane to me, because we're cutting 4.2.18 after 4.3.0, and 4.3.0 -> 4.2.18 is probably not what most clusters want ;).

Reply:
Right...when you have multiple branches you do want something more sophisticated.  We ended up implementing [ref binding](https://github.com/ostreedev/ostree/commit/cc9a0386c46bbeca8f1ce80239eb00ad6d792de8) in libostree, which avoids parsing the version strings.  But I guess for the fully general case of switching branches, you do need something that is parsing them.